File: examples/metasona_example.c
/* SPDX-FileCopyrightText: 2026 MetaSona contributors */
/* SPDX-License-Identifier: Apache-2.0 */

#include <metasona/metasona.h>

#include <math.h>
#include <stdio.h>
#include <stdlib.h>

#define EXAMPLE_PI 3.14159265358979323846264338327950288

static int report_error(const char *operation, ms_status status)
{
    (void)fprintf(
        stderr,
        "%s failed (%d): %s\n",
        operation,
        (int)status,
        ms_status_string(status));
    return EXIT_FAILURE;
}

int main(void)
{
    const size_t sample_count = MS_SAMPLE_RATE_HZ;
    const double rms_pressure_pa = 0.00002 * 1000.0; /* 60 dB SPL. */
    const double peak_pressure_pa = sqrt(2.0) * rms_pressure_pa;
    double *pressure_pa = NULL;
    double specific[MS_BARK_BANDS];
    double loudness_sone = 0.0;
    double sharpness_acum = 0.0;
    double *time_loudness = NULL;
    double *time_specific = NULL;
    size_t frame_count = 0u;
    size_t frame_count_written = 0u;
    size_t index;
    ms_status status;

    pressure_pa = (double *)malloc(sample_count * sizeof(*pressure_pa));
    if (pressure_pa == NULL) {
        (void)fprintf(stderr, "pressure allocation failed\n");
        return EXIT_FAILURE;
    }
    for (index = 0u; index < sample_count; ++index) {
        const double time_s = (double)index / (double)MS_SAMPLE_RATE_HZ;
        pressure_pa[index] = peak_pressure_pa
                           * sin(2.0 * EXAMPLE_PI * 1000.0 * time_s);
    }

    status = ms_loudness_stationary(
        pressure_pa,
        sample_count,
        MS_SAMPLE_RATE_HZ,
        MS_SOUND_FIELD_FREE,
        &loudness_sone,
        specific,
        MS_BARK_BANDS);
    if (status != MS_OK) {
        free(pressure_pa);
        return report_error("ms_loudness_stationary", status);
    }

    status = ms_sharpness_din(
        specific,
        MS_BARK_BANDS,
        &sharpness_acum);
    if (status != MS_OK) {
        free(pressure_pa);
        return report_error("ms_sharpness_din", status);
    }

    status = ms_loudness_time_frame_count(
        sample_count,
        MS_SAMPLE_RATE_HZ,
        &frame_count);
    if (status != MS_OK) {
        free(pressure_pa);
        return report_error("ms_loudness_time_frame_count", status);
    }
    if (frame_count > SIZE_MAX / MS_BARK_BANDS) {
        free(pressure_pa);
        (void)fprintf(stderr, "specific-output size overflow\n");
        return EXIT_FAILURE;
    }
    time_loudness = (double *)malloc(frame_count * sizeof(*time_loudness));
    time_specific = (double *)malloc(
        frame_count * MS_BARK_BANDS * sizeof(*time_specific));
    if (time_loudness == NULL || time_specific == NULL) {
        free(time_specific);
        free(time_loudness);
        free(pressure_pa);
        (void)fprintf(stderr, "time-output allocation failed\n");
        return EXIT_FAILURE;
    }

    status = ms_loudness_time(
        pressure_pa,
        sample_count,
        MS_SAMPLE_RATE_HZ,
        MS_SOUND_FIELD_FREE,
        time_loudness,
        frame_count,
        time_specific,
        frame_count * MS_BARK_BANDS,
        &frame_count_written);
    if (status != MS_OK) {
        free(time_specific);
        free(time_loudness);
        free(pressure_pa);
        return report_error("ms_loudness_time", status);
    }

    (void)printf("native version: %s (ABI %u)\n",
                 ms_version_string(), (unsigned int)ms_abi_version());
    (void)printf("stationary loudness: %.6f sone\n", loudness_sone);
    (void)printf("sharpness: %.6f acum\n", sharpness_acum);
    (void)printf("time-varying frames: %llu (first %.6f sone)\n",
                 (unsigned long long)frame_count_written, time_loudness[0]);

    free(time_specific);
    free(time_loudness);
    free(pressure_pa);
    return EXIT_SUCCESS;
}

File: examples/metasona_example.py
# SPDX-FileCopyrightText: 2026 MetaSona contributors
# MetaSona author: Jiahua Zhang, September 2026.
# SPDX-License-Identifier: Apache-2.0

"""Calculate every public signal metric from calibrated mono pressure."""

from __future__ import annotations

import numpy as np

import metasona as ms


def calibrated_modulated_tone(sample_rate_hz: int, duration_s: float) -> np.ndarray:
    """Return a 1 kHz/70 Hz fully modulated signal at total 60 dB SPL."""
    count = round(sample_rate_hz * duration_s)
    time_s = np.arange(count, dtype=np.float64) / sample_rate_hz
    signal = (1.0 + np.sin(2.0 * np.pi * 70.0 * time_s)) * np.sin(2.0 * np.pi * 1_000.0 * time_s)
    target_rms_pa = 20e-6 * 10 ** (60.0 / 20.0)
    return signal * (target_rms_pa / np.sqrt(np.mean(np.square(signal))))


def main() -> None:
    # 44.1 kHz intentionally demonstrates the single Python resampling boundary.
    sample_rate_hz = 44_100
    pressure_pa = calibrated_modulated_tone(sample_rate_hz, 2.0)

    stationary = ms.stationary_loudness(
        pressure_pa,
        sample_rate_hz,
        sound_field=ms.SoundField.FREE,
    )
    varying = ms.time_varying_loudness(
        pressure_pa,
        sample_rate_hz,
        sound_field="free",
        time_skip_s=0.2,
    )
    roughness = ms.roughness_daniel_weber(
        pressure_pa,
        sample_rate_hz,
        time_skip_s=0.2,
    )
    tonality = ms.tonality_aures(
        pressure_pa,
        sample_rate_hz,
        sound_field="free",
        time_skip_s=0.2,
    )
    sharpness = ms.sharpness_din45692(stationary.specific_loudness_sone_per_bark)

    print(f"native library: {ms.native_version()}")
    print(f"stationary loudness: {stationary.loudness_sone:.6f} sone")
    print(f"sharpness: {sharpness.sharpness_acum:.6f} acum")
    print(
        f"time-varying loudness: {varying.time_s.size} frames, "
        f"mean {np.mean(varying.loudness_sone):.6f} sone"
    )
    print(
        f"roughness: {roughness.time_s.size} frames, "
        f"mean {np.mean(roughness.roughness_asper):.6f} asper"
    )
    print(f"tonality: {tonality.time_s.size} frames, mean {np.mean(tonality.tonality):.6f}")


if __name__ == "__main__":
    main()
